File: tools/pulse_expand.py
#!/usr/bin/env python3
"""Expand a scene's pulse streams into strike cues.

Beat markers from the audio render, turned into light pulses — the Python
copy of the arithmetic in web/src/track_lights.ts (bandStrikes), consuming
the shared per-stream dynamics in pulse_dynamics.py. Split from
gen_esphome.py along the seam that was already documented there: this is
what a pulse stream MEANS; gen_esphome.py is how the device is told.

The same arithmetic lives in web/src/track_lights.ts and is exercised
digit-for-digit by web/test/fuzz_parity.mjs -> tools/fuzz_check.py.
"""

from __future__ import annotations

import math

from pulse_dynamics import (
    PAN_DECISIVE,
    TAKEOVER_COLORS,
    TAKEOVER_HOT,
    drift_base,
    gate_mul,
    gate_note,
    is_accent,
    round3,
    section_gates,
    tempo_decay,
    tempo_factor,
)

WHITE = [1.0, 1.0, 1.0, 1.0]   # default strike colour (lightning)
DEFAULT_DECAY = 0.90           # default strike decay per 16 ms frame


def blend_color(base: list, hot: list | None, vel: float) -> list:
    """color -> color_hot by velocity. Identical in track_lights.ts."""
    if not hot:
        return base
    return [round3(b + (h - b) * vel) for b, h in zip(base, hot)]


def pixels_for(cfg: dict, vel: float) -> str:
    """Velocity picks the strike mask: soft centre, medium scatter, hard all.

    Thresholds are shared with track_lights.ts and gen_previewer.py.
    """
    if not cfg.get("pixels_by_vel"):
        return str(cfg.get("pixels", "all"))
    return "center" if vel < 0.40 else ("scatter" if vel < 0.72 else "all")


def pulse_cues(scene: dict, markers: dict) -> list[dict]:
    """Beat markers from the audio render, turned into light pulses.

    The renderer reports the ACTUAL event times AND loudness of every sound
    (both heartbeat thumps, each whispered word, chord onsets, downbeats), so
    these stay locked to the audio even when a synth jitters its own timing.

    A scene opts in with a `pulse:` list — one stream per synth, each with
    its own colour, intensity and decay, so a heartbeat can be a fast red
    snap while a bell toll is a slow violet bloom. `zones` + `alternate`
    round-robins markers across zones (whispers moving between the towers).

    Dynamics, all per hit, all driven by the marker's velocity:
      colors:        a LIST of colours the stream cycles through hit by hit,
                     so consecutive hits differ in hue, not just brightness.
                     Overrides `color` when present.
      color_hot:     a second colour; each hit blends its base->color_hot by
                     vel, so soft hits sit deep in the hue and hard hits go
                     bright. Keep it saturated: a white-ish hot colour makes
                     every LOUD hit white, and the loud hits are the ones you
                     see — the show reads as "mostly white" (round-3 gauntlet
                     screenshots proved it).
      pixels_by_vel: soft hits touch the centre, medium hits scatter, hard
                     hits take the whole jewel (overrides `pixels`).
      boost_at/boost_targets: a hit at or above `boost_at` spills onto the
                     extra zones too — the big downbeat that lights the castle.
      ms:            strike length for this stream (default 120).

    The same arithmetic lives in web/src/track_lights.ts and
    tools/gen_previewer.py — keep all three in lockstep.
    """
    scene_marks = markers.get(scene["id"], {})
    gates = section_gates(scene)
    out = []
    for cfg in scene.get("pulse") or []:
        beats = scene_marks.get(cfg["synth"], [])
        if not beats:
            print(f"note: scene {scene['id']}: no markers for synth "
                  f"{cfg['synth']!r} — pulse stream skipped")
        zones = cfg.get("zones") or ([cfg["zone"]] if cfg.get("zone") else None)
        # Times arrive in ms here; the tempo maths speaks seconds everywhere.
        factor = tempo_factor([b[0] / 1000.0 for b in beats])
        decay = tempo_decay(cfg.get("decay", DEFAULT_DECAY), factor)
        ms = math.floor(int(cfg.get("ms", 120)) * factor + 0.5)
        vels = [b[1] for b in beats]
        for i, beat in enumerate(beats):
            t, vel = beat[0], beat[1]
            pan = beat[2] if len(beat) > 2 else None
            mul = gate_mul(cfg["synth"], gates, t)
            if mul is None:
                continue                       # gated out by its section (#9)
            targets: list[str] | None
            if zones and cfg.get("alternate"):
                # A decisively panned hit goes to ITS tower (#7); everything
                # else keeps the round-robin movement.
                if (pan is not None and abs(pan) >= PAN_DECISIVE
                        and "towerL" in zones and "towerR" in zones):
                    targets = ["towerL" if pan < 0 else "towerR"]
                else:
                    targets = [zones[i % len(zones)]]
            else:
                targets = list(zones) if zones else None   # None -> all zones
            if (targets and cfg.get("boost_targets")
                    and (vel >= cfg.get("boost_at", 2) or is_accent(vels, i))):
                targets = targets + [z for z in cfg["boost_targets"]
                                     if z not in targets]
            cyc = cfg.get("colors")
            hot = cfg.get("color_hot")
            if cfg.get("takeover") and gate_note(gates, t) == "chorus":
                # #2: in a chorus the castle agrees on one warm family.
                base = TAKEOVER_COLORS[i % len(TAKEOVER_COLORS)]
                hot = TAKEOVER_HOT
            elif cyc and cfg.get("drift"):
                base = drift_base(cyc, i, t)     # #1: hues walk over time
            else:
                base = cyc[i % len(cyc)] if cyc else cfg.get("color", WHITE)
            out.append({"t": t, "op": "strike", "targets": targets,
                        "ms": ms,
                        "intensity": round3(cfg.get("intensity", 0.3) * vel * mul),
                        "color": blend_color(base, hot, vel),
                        "decay": decay,
                        # #10: rise time to peak; 0 keeps the instant slam.
                        "attack": int(cfg.get("attack_ms", 0)),
                        # WHERE on the jewel the pulse lands: a bass thump can
                        # hit the door's centre while highs scatter the rings.
                        "pixels": pixels_for(cfg, vel),
                        "note": cfg["synth"]})
    return out

File: core/src/pulse_expand.rs
//! Expand pulse streams into strike cues — B2 pass 2.
//!
//! Port of `pulse_cues` in `tools/pulse_expand.py` (whose twin is
//! bandStrikes in `web/src/track_lights.ts`), minus the plumbing: the
//! Python walks the scene YAML and the markers file; this takes the
//! streams and section gates it found and does the arithmetic. Every
//! default that lives in a `.get(key, default)` there is a field default
//! here, so the parity corpus pins those numbers as well.

use crate::pulse::{
    PAN_DECISIVE, drift_base, gate_mul, gate_note, is_accent, round3, tempo_decay, tempo_factor,
};

pub const WHITE: [f64; 4] = [1.0, 1.0, 1.0, 1.0];
pub const DEFAULT_DECAY: f64 = 0.90;

/// #2 Chorus takeover family — same table as pulse_dynamics.py.
pub const TAKEOVER_COLORS: [[f64; 4]; 3] = [
    [1.0, 0.55, 0.0, 0.0],  // gold
    [1.0, 0.25, 0.0, 0.05], // flame orange
    [1.0, 0.0, 0.35, 0.0],  // hot rose
];
pub const TAKEOVER_HOT: [f64; 4] = [1.0, 0.75, 0.1, 0.12];

/// One marker hit: time (ms), velocity, optional pan.
#[derive(Clone, Debug)]
pub struct Hit {
    pub t: i64,
    pub vel: f64,
    pub pan: Option<f64>,
}

/// One `pulse:` stream config, defaults matching the Python `.get`s.
#[derive(Clone, Debug, Default)]
pub struct PulseCfg {
    pub synth: String,
    pub zones: Vec<String>, // empty -> all zones (targets None)
    pub alternate: bool,
    pub boost_targets: Vec<String>,
    pub boost_at: Option<f64>, // .get("boost_at", 2)
    pub color: Option<Vec<f64>>,
    pub color_hot: Option<Vec<f64>>,
    pub colors: Vec<Vec<f64>>, // empty -> no cycle
    pub takeover: bool,
    pub drift: bool,
    pub pixels: Option<String>, // .get("pixels", "all")
    pub pixels_by_vel: bool,
    pub intensity: Option<f64>, // .get("intensity", 0.3)
    pub decay: Option<f64>,     // .get("decay", DEFAULT_DECAY)
    pub ms: Option<i64>,        // .get("ms", 120)
    pub attack_ms: i64,
}

/// One expanded strike cue. `targets` None means "all zones".
#[derive(Clone, Debug)]
pub struct Cue {
    pub t: i64,
    pub targets: Option<Vec<String>>,
    pub ms: i64,
    pub intensity: f64,
    pub color: Vec<f64>,
    pub decay: f64,
    pub attack: i64,
    pub pixels: String,
    pub note: String,
}

/// color -> color_hot by velocity. Identical in track_lights.ts.
pub fn blend_color(base: &[f64], hot: Option<&[f64]>, vel: f64) -> Vec<f64> {
    match hot {
        None => base.to_vec(),
        Some(h) => base
            .iter()
            .zip(h.iter())
            .map(|(b, h)| round3(b + (h - b) * vel))
            .collect(),
    }
}

/// Velocity picks the strike mask: soft centre, medium scatter, hard all.
pub fn pixels_for(cfg: &PulseCfg, vel: f64) -> String {
    if !cfg.pixels_by_vel {
        return cfg.pixels.clone().unwrap_or_else(|| "all".to_string());
    }
    if vel < 0.40 {
        "center"
    } else if vel < 0.72 {
        "scatter"
    } else {
        "all"
    }
    .to_string()
}

/// The body of pulse_cues for streams the caller already resolved:
/// (cfg, that synth's beats), plus the scene's section gates.
pub fn pulse_cues(streams: &[(PulseCfg, Vec<Hit>)], gates: &[(i64, String)]) -> Vec<Cue> {
    let mut out = Vec::new();
    for (cfg, beats) in streams {
        let zones = &cfg.zones;
        let factor = tempo_factor(
            &beats
                .iter()
                .map(|b| b.t as f64 / 1000.0)
                .collect::<Vec<_>>(),
        );
        let decay = tempo_decay(cfg.decay.unwrap_or(DEFAULT_DECAY), factor);
        let ms = (cfg.ms.unwrap_or(120) as f64 * factor + 0.5).floor() as i64;
        let vels: Vec<f64> = beats.iter().map(|b| b.vel).collect();
        for (i, beat) in beats.iter().enumerate() {
            let (t, vel, pan) = (beat.t, beat.vel, beat.pan);
            let Some(mul) = gate_mul(&cfg.synth, gates, t) else {
                continue; // gated out by its section (#9)
            };
            let mut targets: Option<Vec<String>> = if !zones.is_empty() && cfg.alternate {
                // A decisively panned hit goes to ITS tower (#7).
                let both =
                    zones.iter().any(|z| z == "towerL") && zones.iter().any(|z| z == "towerR");
                match pan {
                    Some(p) if p.abs() >= PAN_DECISIVE && both => {
                        Some(vec![if p < 0.0 { "towerL" } else { "towerR" }.to_string()])
                    }
                    _ => Some(vec![zones[i % zones.len()].clone()]),
                }
            } else if zones.is_empty() {
                None // all zones
            } else {
                Some(zones.clone())
            };
            if let Some(tg) = &mut targets {
                if !cfg.boost_targets.is_empty()
                    && (vel >= cfg.boost_at.unwrap_or(2.0) || is_accent(&vels, i))
                {
                    for z in &cfg.boost_targets {
                        if !tg.contains(z) {
                            tg.push(z.clone());
                        }
                    }
                }
            }
            let mut hot = cfg.color_hot.as_deref();
            let base: Vec<f64> = if cfg.takeover && gate_note(gates, t) == Some("chorus") {
                hot = Some(&TAKEOVER_HOT);
                TAKEOVER_COLORS[i % TAKEOVER_COLORS.len()].to_vec()
            } else if !cfg.colors.is_empty() && cfg.drift {
                drift_base(&cfg.colors, i, t)
            } else if !cfg.colors.is_empty() {
                cfg.colors[i % cfg.colors.len()].clone()
            } else {
                cfg.color.clone().unwrap_or_else(|| WHITE.to_vec())
            };
            out.push(Cue {
                t,
                targets,
                ms,
                intensity: round3(cfg.intensity.unwrap_or(0.3) * vel * mul),
                color: blend_color(&base, hot, vel),
                decay,
                attack: cfg.attack_ms,
                pixels: pixels_for(cfg, vel),
                note: cfg.synth.clone(),
            });
        }
    }
    out
}
